### ISSUE TITLE
Events List does not honor stripe_version kwarg

### ISSUE THREAD

**Issue by KAllan357** [stripe/stripe-python#622]
Hello, I am observing some behavior that seems incorrect and I'm looking for some guidance.

```
$ python --version
Python 3.5.6
$ pip list
Package           Version    Location            
----------------- ---------- --------------------
... 
stripe            2.37.2
...
```

What I've noticed is that this SDK may not be sending the stripe_version header when using the `stripe.Event.list()` function, which is specifically causing an issue for an Event relating to Invoices' Line Item's `period` field.

Here's some sample code:

```
import stripe                                                                                                                                                                                                       

stripe.api_key = '<secret>'

stripe.api_version = '2018-09-24'

date_window_start = 1567728000
date_window_end = 1568332800

for e in stripe.Event.list(**{"limit": 100,
                              "type": "invoice.*",
                              "stripe_account": '<account>',
                              "created[gte]": date_window_start,
                              "created[lt]": date_window_end}).auto_paging_iter():
    if e.get('data').get('object').get('id') in ['<affected invoice>', '<affected invoice>', '<affected invoice>']:
        print(e.get('id'))
        print(e.get('api_version'))
        period_data = [l.get('period') for l in e.get('data').get('object').get('lines').to_dict().get('data')]
        print(period_data)
```

The output is as follows:

```
<event id>
2014-11-20
[<StripeObject at 0x7f716ba94a48> JSON: {
  "end": 1567896206,
  "start": 1567896206
}, <StripeObject at 0x7f716ba94bd8> JSON: {
  "end": 1567896206,
  "start": 1567896206
}]
```
Specifically adding another kwarg `stripe_version` seems to have no affect on the above behavior.

Ultimately, the period's start and end value should not match. This seems to correspond to the API version in the response being a version before the changelog entry [here](https://stripe.com/docs/upgrades#2015-04-07) which seems to document a change in this behavior.

If you make the same request in curl, the API honors the header and returns data that appears correct:

```
curl --globoff -v -H "Stripe-Version: 2018-09-24" -u <secret>: 
"https://api.stripe.com/v1/events?type=invoice.%2A&starting_after=evt_1FGGDgH1INVpfABmlEYR9Mts&created[gte]=1567728000&limit=100&created[lt]=1568332800"

     "id": "evt_1234",
      "object": "event",
      "api_version": "2019-05-16",
      "data": { "object": {"lines": { "data": [{"period": {"end": 1567900915, "start": 1567796676}}]}}
```

**Comment by remi-stripe** [comment]
Events are a bit different from most objects in our API. They are calculated at the time of creation based on the API version set on the webhook endpoint or the account and they are "pinned" to that version forever. You can not fetch a given event and render its `data` hash based on another API version, this just will not work. Only the top-level properties would re-render (which is confusing).

**Comment by KAllan357** [comment]
@remi-stripe Thanks for the response. If I understand correctly, isn't what I'm seeing? Two different API versions returning different data.

I use this SDK and the events come back with an old `api_version` and different data compared to a curl request which returns a newer `api_version`. The Event's ID is exactly the same in both cases - the data inside the event is what changes.

**Comment by remi-stripe** [comment]
@KAllan357 I'm fairly certain it's impossible for anything inside `data` to change. The top-level properties would change based on the API version but what is inside `data` should not.

If you see otherwise, would you be able to provide an example of an event id where you get a different JSON  in `data` when fetching the same event with different API versions?

**Comment by KAllan357** [comment]
Sure, I have quite a few examples:

evt_1FGwFJH1INVpfABmJunyl4Q9
evt_1FGCdmH1INVpfABmnpx5ZhY9
evt_1FGCfAH1INVpfABmTR1YbOPQ
evt_1FGCfAH1INVpfABmaf8g2WAv
evt_1FGCfBH1INVpfABmdMKfkRVM

There are more, but that should be enough. My code above is able to iterate events between the two timestamps, see the invoice line item's period being the same.

A corresponding curl to each event ID shows different period data.

**Comment by remi-stripe** [comment]
@KAllan357 I'm really sorry but I'm still having quite a lot of troubles grasping what you are describing exactly here. It seems there are 2 separate issues at hand but we're discussing a different one together.

The shape if the event in `data` should never change. Re-reading the exchange I don't think the shape is changing so that part is likely just working as expected. If that's wrong, would you be able to provide clear output of the calls with 2 different API version when the JSON inside `data` changes for the same event based on the API version you pass in the call to List Events?

This leaves me to think the problem you're focusing on is that the period start and end are identical.  This can happen in various cases such as on the first invoice of a Subscription for example. I looked at the event `evt_1FGwFJH1INVpfABmJunyl4Q9` you mentioned and that one has different periods. This is likely due to this API version change for [2015-04-07](https://stripe.com/docs/upgrades#2015-04-07):
> Updates the period[end] property on proration invoice line items to reflect the subscription’s current_period_end property when the update and proration was made. A proration invoice line item’s period[start] and period[end] properties now represent the prorated adjustment interval. Previously, period[end] marked the time at which the proration was made, and was the same as period[start].

This event likely was generated with an API version before that one and so is snapshot-ed to a time where the period start and end were identical.

**Comment by KAllan357** [comment]
@remi-stripe no problem, this is somewhat difficult to explain. You're right, I'm not claiming the shape as different, only the data. Despite the API version changing, the shape remains the same.

Yes, the period start and end being identical is the problem I'm looking into. What strikes me as odd, is that I can use this SDK to list events, find the one you pasted and I see them with identical periods:

```
evt_1FGwFJH1INVpfABmJunyl4Q9
2014-11-20
[<StripeObject at 0x7f20017c3e08> JSON: {
  "end": 1568071496,
  "start": 1568071496
}, <StripeObject at 0x7f20017c3f98> JSON: {
  "end": 1568071496,
  "start": 1568071496
}]
```

Or grab the Event by its ID, and observe different periods:

```
curl https://api.stripe.com/v1/events/evt_1FGwFJH1INVpfABmJunyl4Q9 | jq .data.object.lines.data[].period

{
  "end": 1570258860,
  "start": 1568071496
}
{
  "end": 1570258860,
  "start": 1568071496
}
```

**Comment by remi-stripe** [comment]
@KAllan357 But that likely just means this event was generated with an API version before the one I linked. And once the event is created, it will never mutate, even if you fetch it with a different API version. It's really hard to help here without a lot more concrete details.

My guess is that you are not fetching the event the same way in both cases and we would need some really specific information on all the requests you are making to clarify if for example they use Connect. I'd recommend reaching out to our support team. Feel free to mention my name and the link to the issue and we can investigate deeper.

**Comment by remi-stripe** [comment]
Circling back after we worked on this question internally. We were able to confirm that the issue was due to the older API version used by the integration in that case and there is no bug in the API or the stripe-python library.

The issue is that when retrieving data on the account, the information was present as expected. The third-party platform used though is on an API version from 2014. All the events created on the account for this platform used that older API version. The data is then quite different as a lot changed in our API in the past 5 years.

The right path forward here would be for the third-party integration to upgrade their API version to the latest one to be able to ingest data formatted based on our latest improvements such as having the right period information for line items.
